src/Reflection/ReflectionFunctionAbstract.php

```diff
@@ -619,4 +619,17 @@ public function getReturnStatementsAst()
 
         return $visitor->getReturnNodes();
     }
+
+    private function findTypeFromAst($namespace, $locatedSource, $type)
+    {
+        $objectType = (new FindTypeFromAst())->__invoke(
+            $this->reflector->getContextFactory()->createForNamespace(
```

Comment:
> 3) potential improvement of performance (good).

So, do I understand corrcetly, we instantiate a `[Cached]ContextFactory` for each `Reflection[Class|Function]`. This would be good I think - in most cases it would not benefit to have a "shared" cache between reflection classes as people generally define a single class per file, for functions the performance benefit would be less, but we can still benefit from not parsing the AST for each function argument (and later the same for variable type inference).

We could default to `[InMemory]CachedContextFactory` and potentially still allow injection of the factory which would be compatible with future refactoring.

Did I understand it right?